crates/blockchain-tree/src/block_buffer.rs

```diff
@@ -47,30 +43,47 @@ impl BlockBuffer {
         Self {
             blocks: Default::default(),
             parent_to_child: Default::default(),
-            hash_to_num: Default::default(),
+            earliest_blocks: Default::default(),
             lru: LruCache::new(NonZeroUsize::new(limit).unwrap()),
             metrics: Default::default(),
         }
     }
 
+    /// Return reference to buffered blocks
+    pub fn blocks(&self) -> &HashMap<BlockHash, SealedBlockWithSenders> {
+        &self.blocks
+    }
+
+    /// Return reference to the requested block.
+    pub fn block(&self, hash: &BlockHash) -> Option<&SealedBlockWithSenders> {
+        self.blocks.get(hash)
+    }
+
+    /// Return a reference to the lowest ancestor of the given block in the buffer.
+    pub fn lowest_ancestor(&self, hash: &BlockHash) -> Option<&SealedBlockWithSenders> {
+        let mut current_block = self.blocks.get(hash)?;
+        while let Some(parent) = self.blocks.get(&current_block.parent_hash) {
+            current_block = parent;
+        }
+        Some(current_block)
+    }
+
     /// Insert a correct block inside the buffer.
     pub fn insert_block(&mut self, block: SealedBlockWithSenders) {
-        let num_hash = block.num_hash();
+        let hash = block.hash;
 
-        self.parent_to_child.entry(block.parent_hash).or_default().insert(block.num_hash());
-        self.hash_to_num.insert(block.hash, block.number);
-        self.blocks.entry(block.number).or_default().insert(block.hash, block);
+        self.parent_to_child.entry(block.parent_hash).or_default().insert(hash);
+        self.earliest_blocks.entry(block.number).or_default().insert(hash); // TODO: fix removal.
```

Comment:
what's to fix here?
can we fix?

Reply:
oops, was just a reminder, already fixed